### PR TITLE
[SPARK-30617][SQL] Stop check values of spark.sql.catalogImplementation to improve expansibility

### PR DESCRIPTION
**Pull request by WeiWenda**
### What changes were proposed in this pull request?

When user config spark.sql.catalogImplementation with value not in in-memory/hive, check if below properties is configured. If configured then instantiate SessionState with provided Class, or else throw Exception as usual. 
```
spark.sql.catalogImplementation.[value of spark.sql.catalogImplementation].builder
spark.sql.catalogImplementation.[value of spark.sql.catalogImplementation].externalCatalog
```
For example:
```
spark.sql.catalogImplementation = qihoo
spark.sql.catalogImplementation.qihoo.builder = org.apache.spark.sql.qihoo.QihooSessionStateBuilder
spark.sql.catalogImplementation.qihoo.externalCatalog = org.apache.spark.sql.qihoo.QihooExternalCatalog
```

### Why are the changes needed?
We have implemented a complex ExternalCatalog which is used for retrieving multi isomerism database's metadata(sush as elasticsearch、postgresql), so that we can make a mixture query between hive and our online data. But as spark require that value of spark.sql.catalogImplementation must be one of in-memory/hive, we have to modify SparkSession and rebuild spark to make our project work.
Finally, we hope spark removing above restriction, so that it's will be much easier to let us keep pace with new spark version. Thanks!

### Does this PR introduce any user-facing change?
no
### How was this patch tested?
no

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala [hunk 1]
```diff
@@ -36,9 +36,13 @@ object StaticSQLConf {
     .createWithDefault(Utils.resolveURI("spark-warehouse").toString)
 
   val CATALOG_IMPLEMENTATION = buildStaticConf("spark.sql.catalogImplementation")
+    .doc("Spark built-in implementation is in-memory and hive," +
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: nit. `,"` -> `, "`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala [hunk 2]
```diff
@@ -36,9 +36,13 @@ object StaticSQLConf {
...
+      "when set to yourImpl(can be any words), " +
+      "you need also provide following configuration to make everything going" +
+      "spark.sql.catalogImplementation.yourImpl.builder" +
+      "spark.sql.catalogImplementation.yourImpl.externalCatalog")
     .internal()
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: Technically, this is `.internal()` configuration. This PR seems to propose make this external conf.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 3]
```diff
@@ -1026,6 +1026,12 @@ object SparkSession extends Logging {
     conf.get(CATALOG_IMPLEMENTATION) match {
       case "hive" => HIVE_SESSION_STATE_BUILDER_CLASS_NAME
       case "in-memory" => classOf[SessionStateBuilder].getCanonicalName
+      case other => conf.getOption(s"spark.sql.catalogImplementation.$other.builder")
+          .getOrElse {
+        throw new IllegalArgumentException(
+          "You need to configure spark.sql.catalogImplementation.xx.builder when xx configured by" +
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: The following will be better.
```scala
- "You need to configure spark.sql.catalogImplementation.xx.builder when xx configured by" +
+ s"You need to configure spark.sql.catalogImplementation.$other.builder when $other configured by" +
```

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 4]
```diff
@@ -217,6 +217,12 @@ object SharedState extends Logging {
     conf.get(CATALOG_IMPLEMENTATION) match {
       case "hive" => HIVE_EXTERNAL_CATALOG_CLASS_NAME
       case "in-memory" => classOf[InMemoryCatalog].getCanonicalName
+      case other => conf.getOption(s"spark.sql.catalogImplementation.$other.externalCatalog")
+          .getOrElse {
+        throw new IllegalArgumentException(
+          "You need to configure spark.sql.catalogImplementation.xx.externalCatalog when xx " +
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala]: ditto.

**Comment by dongjoon-hyun** [comment]
Technically, there was an existing effort on this.
- [[SPARK-17767][SQL] Support user-provided ExternalCatalog](https://github.com/apache/spark/pull/15336)

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a PR, @WeiWenda .

The recent community decision was using new `CatalogPlugin` for this kind of needs.

Since this is requested again, I'll ping the senior PMCs once more.

Hi, @rxin , @gatorsmile , @cloud-fan .
Can we extend `spark.sql.catalogImplementation` in `3.0.0`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala (hunk 2 above)
Comment (WeiWenda) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: I find that internal didn't effect anything. So I have removed .internal() in latest commit.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala (hunk 1 above)
Comment (WeiWenda) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: fixed

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 3 above)
Comment (WeiWenda) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: fixed

**Comment by WeiWenda** [comment]
@dongjoon-hyun Thanks for your help. SPARK-17767 and SPARK-19632's solution is so old that cannot be used in spark 3.0.0 anymore. 
> The recent community decision was using new CatalogPlugin for this kind of needs.
If there still has a long way to go with new CatalogPlugin, maybe cancel checkValues of `spark.sql. catalogImplementation` is lowest-cost way to consider user-defined catalog.

**Comment by dongjoon-hyun** [comment]
I don't think so.
> SPARK-19632's solution is so old that cannot be used in spark 3.0.0 anymore.

`CatalogPlugin` is already available from 3.0.0-preview via https://issues.apache.org/jira/browse/SPARK-24252 .
> If there still has a long way to go with new CatalogPlugin

**Comment by WeiWenda** [comment]
Yeah, i noticed your SPARK-17767, you are foresight and responsible. I just look at https://github.com/apache/spark/pull/23915, it can solve our problem more gently. Is spark 3.0.0 will release in recent month?

**Comment by dongjoon-hyun** [comment]
According to the plan, we will proceed `feature freeze` first by creating `branch-3.0` in this month.
- https://spark.apache.org/versioning-policy.html

Then, QA period will be started. For major version, it takes some time.

In any way, your patch cannot be released faster than that because new feature is not backported to `branch-2.4`. `2.4.5 RC1` vote was held last week and we are receiving some more correctness patches. `2.4.5 RC2` vote will start soon.

**Comment by dongjoon-hyun** [comment]
BTW, if you agree with `CatalogPlugin` in 3.0.0, could you close the PR and JIRA please?

**Comment by WeiWenda** [comment]
> BTW, if you agree with `CatalogPlugin` in 3.0.0, could you close the PR and JIRA please?

All right, we will waiting for spark 3.0.0. Thank you anyway.

**Comment by dongjoon-hyun** [comment]
Thank you, @WeiWenda .

**Pull request closed without merging**